include/cutlass/fp8_to_fp16.h

```diff
@@ -117,12 +121,133 @@ static inline ushort16 E4M3_to_FP16_chunk16(uchar16 xin) {
+  //   (static_cast<uint16_t>((pSrc[i]))) << 8;
+  // }
+  // The root-cause is unknown, but private memory use is seen in that case.
+  // We're using a workaround that doesn't use private memory.
   CUTLASS_PRAGMA_UNROLL
   for (int i = 0; i < N; i++) {
     xout[i] = (static_cast<uint16_t>(xin[i])) << 8;
   }
 }
+
+template <typename ElementA, class EngineIn, class EngineOut, class LayoutIn,
+          class LayoutOut, class... Ts>
+CUTLASS_DEVICE void
+convert_FP8_to_FP16(cute::Tensor<EngineIn, LayoutIn> const &in,
+                    cute::Tensor<EngineOut, LayoutOut> &out) {
+
+  static_assert(cute::is_rmem<EngineIn>::value,
+                "Input tensor for A conversion must come from registers");
+  static_assert(cute::is_rmem<EngineOut>::value,
+                "Output tensor for A conversion must come from registers");
+  static_assert(cute::cosize_v<LayoutIn> == cute::cosize_v<LayoutOut>);
+  static_assert(cute::size_v<LayoutIn> == cute::cosize_v<LayoutIn>);
+  static_assert(cute::size_v<LayoutOut> == cute::cosize_v<LayoutOut>);
+
+  using SrcType = typename EngineIn::value_type;
+  using DstType = typename EngineOut::value_type;
+
+  static_assert(std::is_same_v<SrcType, uint8_t>,
+                "Expected fp8 (E4M3) input as uint8_t");
+  static_assert(std::is_same_v<DstType, cute::half_t>,
+                "Expected fp16 output as half_t");
+
+  auto const &src = in(cute::_, cute::_, cute::_);
+  auto const &dst = out(cute::_, cute::_, cute::_);
+
+  SrcType const *pSrc = src.data();
+  DstType *pDst = dst.data();
+
+  constexpr int num_elements = decltype(size(src))::value;
+
+  // TODO(Codeplay): Move conversion to NumericArrayConverter
+  if constexpr (std::is_same_v<ElementA, cute::float_e5m2_t>) {
+    // May convert four FP8 elements at a time
+    constexpr bool use_faster_conversion = num_elements >= 8;
+    using src_dtype =
+        std::conditional_t<use_faster_conversion, uint32_t, uint8_t>;
+    using dst_dtype =
+        std::conditional_t<use_faster_conversion, uint32_t, uint16_t>;
+    using SrcArray =
+        std::conditional_t<use_faster_conversion,
+                           cutlass::Array<src_dtype, num_elements / 4>,
+                           cutlass::Array<src_dtype, num_elements>>;
+    using DstArray =
+        std::conditional_t<use_faster_conversion,
+                           cutlass::Array<dst_dtype, num_elements / 2>,
+                           cutlass::Array<dst_dtype, num_elements>>;
+    SrcArray const *pSrcArr = reinterpret_cast<SrcArray const *>(pSrc);
+    DstArray *pDstArr = reinterpret_cast<DstArray *>(pDst);
+    if constexpr (use_faster_conversion) {
+      // convert 4 FP8 elements at a time
+      E5M2_to_FP16<num_elements>(*pSrcArr, *pDstArr);
+    } else {
+      E5M2_to_FP16<num_elements>(*pSrcArr, *pDstArr);
+    }
+  } else {
+    // E4M3 -> FP16 conversion
+    constexpr int chunk_size = 16;
+    constexpr int iters = num_elements / chunk_size;
```

Comment:
Do you need to assert this is divisible?

Reply:
Will do, thanks!